h2/src/test/org/h2/test/store/TestMVStore.java

```diff
@@ -1361,9 +1373,16 @@ private void testVersion() {
         assertEquals("Hi", m.get("1"));
         assertEquals(null, m.get("2"));
 
+        // This test tries to cast in bronze some peculiar behaviour,
+        // which is rather implementation artifact then intentional.
+        // Once store is closed, only one single version of the data
+        // will exists upon re-opening - the latest.
+        // I hope nobody relies on this "multi-versioning".
+/*
```

Comment:
Ok, it seems I've gone too far here. MV is a feature, of course. What I was trying to avoid is to advertise / publish somewhot confusing and inconsistent behavior. 

@thomasmueller, since I've got your attention here, can we get some clarifications about version numbers and related API?
When store is open for the first time it has version 0, right? If I would create a map, put some data, call commit() (TestMVStore.testExampleMvcc()), and later call m.openVersion(0), returned R/O map would be a snapshot of the state right before the first commit. On the other hand, call s.rollbackTo(0) willrevert to the state when store was just open. Quite confusing, huh? Even worse, m.openVersion() might change behavior, if store was closed/reopened, or more commits were put in between! For instance, if in aforementioned testExampleMvcc, you insert two extra commits (which should not change program semantics at all):
`        map.remove(2);

       s.commit();
       map.put(3, "Dummy");
       s.commit();

        // access the old data (before the commit)
        MVMap<Integer, String> oldMap =
                          map.openVersion(oldVersion);`

test will fail with "IllegalArgumentException: Unknown version 0". That's because this time it fetches the old version not from in-memory list, but from disk and now it's behavior (in terms of version numbers) matches to one of s.rollback() - go back to the "beginning" of the mentioned version, not to the "end" of it ("beginning" of the next version).
So the question is: what's the desired behavior?

Reply:
Yes, I know, the current behavior is unfortunately strange. I also have lots of problems in this area, *specially* if in-memory versions _and_ persistent versions are involved. I'm completely OK with slight changes in the meaning of version numbers. It would be nice if things are more consistent, but I know it's hard to make it consistent. If that's not possible, that's fine. 

But in general, I think it would be good if it's possible to roll back to an old version, both for the in-memory and for the persistent case (including after re-opening). The exact meaning of a version number is not fully set in stone (even though it would be good to document it).

Other changes I'm completely OK. For example, right now, each maps individually remembers a list of old versions (roots actually). It would be better if each *store* remembers a list of old versions (roots of all maps). But probably maybe that would require too many changes.

Reply:
Don't take me wrong, I am not against ability to roll back to an old persistent version, and in this change I've tried hard to preserve this functionality. It just sometimes hard to preserve inconsistency, as it was the case with that unit test.
So, which way you prefer it to work? I personally inclined to the following:
+ s.rollbackTo(X) stays the way it is now - back to state right after call to commit() which returned X
+ m.openVersion(X) should consistently open the map as it was right after call to commit() which returned X, and return null, if map hasn't existed at that time.

Second item will change the existing behavior (most of the time), by shifting version by 1.

As far as you last comment about store remembering a list of old versions - how it is different from what we have now? Is your concern is about atomicity of that save operation and coherent picture? Than I agree - this one is hard to achieve, but we are pretty close.

Reply:
I think, I fixed that one with this change, and TestMVStore.testRemoveMapRollback() should demonstrate it. When map is removed, only name-related entry is deleted from meta-map, it is still accessible by id, until store version where map was removed is disposed of. At this time reference from the store and root location in meta-map are removed.

One of the coherency problems still outstanding is as follows:
+ thread1: commit started, store version already increased
+ thread2: map created (in new version already)
+ thread2: user updates some other map
+ thread1: change to that map picked up and saved as part of a previous version.
+ later user communicates version returned from commit from thread1 to thread2 and rolls back to this commit - new map is not there, but change to other map is 

Solution would be to collect all roots atomically, but it's more involved and hurts concurrency if done naively...